docs/proposals/client-package-structure.md

```diff
@@ -0,0 +1,310 @@
+future, if we have binary-only APIs (unlikely?), we can add that to the
+discovery information and construct an appropriate dynamic Codec.
+
+Action items:
+* A rudimentary version of this exists in kubectl's builder. It needs to be
+  moved to a more general place.
+* Produce a useful 'Unstructured' runtime.Object, which allows for easy
+  Object/ListMeta introspection.
+
+### High-level: Client Sets
+
+Because there will be multiple groups with multiple versions, we will provide an
+aggregation layer that combines multiple typed clients in a single object.
+
+We do this to:
+* Deliver a concrete thing for users to consume, construct, and pass around. We
+  don't want people making 10 typed clients and making a random system to keep
+  track of them.
+* Constrain the testing matrix. Users can generate a client set at their whim
+  against their cluster, but we need to make guarantees that the clients we
+  shipped with v1.X.0 will work with v1.X+1.0, and vice versa. That's not
+  practical unless we "bless" a particular version of each API group and ship an
+  official client set with earch release. (If the server supports 15 groups with
+  2 versions each, that's 2^15 different possible client sets. We don't want to
+  test all of them.)
+
+A client set is generated into its own package. The generator will take the list
+of group/versions to be included. Only one version from each group will be in
+the client set.
+
+A client set is constructable at runtime from either a ClientMux or a transport
+config (for easy one-stop-shopping).
+
+An example:
+
+```go
+import (
+  api_v1 "k8s.io/kubernetes/pkg/client/typed/generated/v1"
+  ext_v1beta1 "k8s.io/kubernetes/pkg/client/typed/generated/extensions/v1beta1"
+  net_v1beta1 "k8s.io/kubernetes/pkg/client/typed/generated/net/v1beta1"
+  "k8s.io/kubernetes/pkg/client/typed/dynamic"
+)
+
+type Client interface {
+  API() api_v1.Client
+  Extensions() ext_v1beta1.Client
+  Net() net_v1beta1.Client
+  // ... other typed clients here.
+
+  // Included in every set
+  Discovery() discovery.Client
+  GroupVersion(group, version string) dynamic.Client
+}
+```
+
+Note that a particular version is chosen for each group. It is a general rule
+for our API structure that no client need care about more than one version of
+each group at a time.
+
+This is the primary deliverable that people would consume. It is also generated.
+
+Action items:
+* This needs to be built. It will replace the ClientInterface that everyone
+  passes around right now.
+
+## Package Structure
+
+```
+pkg/client/
+----------/transport/     # transport & associated config
+----------/restclient/
+----------/clientmux/
+----------/typed/
+----------------/discovery/
+----------------/generated/
+--------------------------/<group>/
+----------------------------------/<version>/
+--------------------------------------------/<resource>.go
+----------------/dynamic/
+----------/clientsets/
+---------------------/release-1.1/
```

Comment:
Thinking more about this, am wondering if we really need to duplicate it here in HEAD.
The client will be there in the corresponding release branch.

How will cherrypicks in release branch work? Will we do the cherrypick, generate the client in release branch and then copy it back here?